kong/tools/queue.lua

```diff
@@ -0,0 +1,369 @@
+      break
+    end
+
+    if not err then
+      self:log(ERR, "handler returned falsy value but no error information")
+    end
+
+    if (now() - start_time) > self.max_retry_time then
+      self:log(
+        ERR,
+        "could not send entries, giving up after %d retries.  %d queue entries were lost",
+        retry_count, entry_count)
+      break
+    end
+
+    self:log(WARN, "handler could not process entries: %s", tostring(err))
+    retry_count = retry_count + 1
+    self.last_used = now()
+    ngx.sleep(math.min(self.max_retry_delay, (retry_count * retry_count) * 0.01))
+  end
+
+  for _ = 1, entry_count do
+    self:delete_frontmost_entry()
+  end
+  if self.queue_full then
+    self:log(INFO, 'queue resumed processing')
+    self.queue_full = false
+  end
+end
+
+
+function Queue.get_params(config)
+  local key = config.__key__
+  local queue = unpack({config.queue or {}})
+  if config.retry_count then
+    ngx.log(ngx.WARN, string.format(
+      "deprecated `retry_count` parameter in plugin %s ignored",
+      key))
+  end
+  if config.queue_size then
+    ngx.log(ngx.WARN, string.format(
+      "deprecated `queue_size` parameter in plugin %s converted to `queue.batch_max_size`",
+      key))
+    queue.batch_max_size = config.queue_size
+  end
+  if config.flush_timeout then
+    ngx.log(ngx.WARN, string.format(
+      "deprecated `flush_timeout` parameter in plugin %s converted to `queue.max_delay`",
+      key))
+    queue.max_delay = config.flush_timeout
+  end
+  return queue
+end
+
+
+-------------------------------------------------------------------------------
+-- Drain the queue, for testing purposes.
+-- @param self Queue
+function Queue:drain()
+  self.running = false
+  while #self.queue > 0 do
+    self:process_once(0.01)
+  end
+end
+
+
+-------------------------------------------------------------------------------
+-- Add entry to the queue
+-- @param conf plugin configuration of the plugin instance that caused the item to be queued
+-- @param entry the value included in the queue. It can be any Lua value besides nil.
+-- @return true, or nil and an error message.
+function Queue:add(data)
+  -- FIXME check argument types
+  if data == nil then
+    return nil, "entry must be a non-nil Lua value"
+  end
+
+  if #self.queue >= self.capacity * 0.9 then
+    if not self.warned then
+      self:log(WARN, 'queue at 90% capacity')
+      self.warned = true
+    end
+  else
+    self.warned = nil
+  end
+
+  if #self.queue == self.capacity then
+    if not self.queue_full then
+      self.queue_full = true
+      self:log(ERR, "queue full, dropping old entries until processing is successful again")
+    end
+    self:delete_frontmost_entry()
+  end
+
+  if self.string_capacity then
+    if type(data) ~= "string" then
+      self:log(ERR, "queuing non-string data to a queue that has queue.string_capacity set, capacity monitoring will not be correct")
+    else
+      if #data > self.string_capacity then
```

Comment:
add `local data_len = #data` may be better.

Reply:
Introducing extra names creates more cognitive load on the person who is reviewing the code.  For performance, it does not matter because the second use of #data happens inside of the logging statement, which so expensive that the extra indirection into the table is insignificant.